pkg/server/extension/extension.go

```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2023 Intel Corporation
+package extension
+
+import (
+	"log"
+
+	"github.com/opiproject/opi-spdk-bridge/pkg/server"
+)
+
+/*
+This function provides the ability to create an extension for xPU by exposing
+default SPDK JSON RPC interface, but requiring adjustments only for some of
+the calls.
+
+To add an extension, a user needs to create a dedicated struct which is
+leveraging Go embedding capabilities for server.OpiServer interface e.g.
+```
+
+	type myXpuServer struct  {
+		server.OpiServer
+		muField1 string
+		muField2 string
+	}
+
+```
+
+and provide customizations for required calls e.g.
+```
+
+	func (s *myXpuServer) CreateVirtioBlk(
+		ctx context.Context,
+		in *pb.CreateVirtioBlkRequest) (*pb.VirtioBlk, error) {
+
+		// do something specific for this xPU
+	}
+
+```
+
+Here, the rest of the calls rely on the default opi-spdk-bridge implementation,
+but CreateVirtioBlk will have its own behavior.
+
+To instantiate myXpuServer, it is needed to embed baseSpdkServer and return the
+instance from Extend function
+```
+
+	return myXpuServer {
+		baseSpdkServer, // embed the default bridge
+		"my param 1",
+		"my param 2"
+	}, nil
+
+```
+*/
+func Extend(baseSpdkServer server.OpiServer) (server.OpiServer, error) {
+	log.Println("No extension is applied.")
+	return baseSpdkServer, nil
+}
```

Comment:
```suggestion

type Extension func(baseSpdkServer server.OpiServer) (server.OpiServer, error)

var extensions []Extension

func RegisterExtension(extension Extension) {
    extensions = append(extensions, extension)
}

func Extend(baseSpdkServer server.OpiServer) (server.OpiServer, error) {
    for _, extension := range extensions {
       s, err := extension(baseSpdkServer)
       if err != nil {
           return nil, err
       }
       baseSpdkServer = s
    }
    return baseSpdkServer
}
```

Reply:
I left out Godoc comments for the sake of brevity.

It's worth calling out here that the extension itself doesn't need to be in this directory. This might be a path towards consuming the code in this repo as an imported package, instead of having to maintain a fork of it.

Reply:
Updated the code and and the comment how to use extensions